Port Truncated Normal and Wald Distributions to V4

matteo-pallini opened the pull request:
Port Truncated Normal and Wald to V4 as per #4686 guidelines

Still need to do/check the followings:

- TruncatedNormal
+ [x] Need to investigate why `pymc3.tests.test_model.TestValueGradFunction.test_aesara_switch_broadcast_edge_cases_2` is failing
+ [x] Is it fine to rewrite a new RV? according to the issue there should be one already, but I couldn't find it 
+ [x] Is it fine to pass `transform` as argument in `dist` ?
+ [x] Is `_defaultval` deprecated? I haven't been able to find any use of it

- Wald
+ [x] Refactor as per [guidelines](https://github.com/pymc-devs/pymc3/issues/4686#issuecomment-839541040)
+ [ ] Investigate why `pymc3.tests.test_distributions_random.TestWaldAlpha` is failing.

matteo-pallini committed:
Refactor Truncated Normal

ricardoV94 commented:
> Is it fine to rewrite a new RV? according to the issue there should be one already, but I couldn't find it

I might have confused it with the truncexpon

ricardoV94 commented on pymc3/distributions/continuous.py:
```diff
@@ -723,42 +686,69 @@ def logp(self, value):
         return ["mu", "sigma", "lower", "upper"]
 
 
+def _truncated_normal_prepare_mu_sigma_params(mu, sigma, tau, sd):
```
I think this would be better directly in the dist classmethod. This sort of initialization logic is what that method is exactly for, so it's okay to have it there.

It shouldn't be needed again in logp.

ricardoV94 commented on pymc3/distributions/continuous.py:
```diff
@@ -723,42 +686,69 @@ def logp(self, value):
+    return mu, sigma
+
+
+def _truncated_normal_normalization(mu, sigma, lower, upper):
```
This can go directly in the logp method I think. Is it used anywhere else? Or at least a classmethod to keep things more organized

ricardoV94 commented on pymc3/distributions/continuous.py:
```diff
@@ -723,42 +686,69 @@ def logp(self, value):
+            bounds.append(value >= lower)
+        if upper_check is not None:
+            bounds.append(value <= upper)
+        print([e.eval() for e in bounds])
```
print statement

matteo-pallini replied:
Ops, thanks

ricardoV94 commented on pymc3/distributions/continuous.py:
```diff
@@ -723,42 +686,69 @@ def logp(self, value):
+        lower, lower_check, upper, upper_check = _truncated_normal_prepare_lower_and_upper(
+            lower, upper
+        )
+        print(lower.eval())
```
print statement

matteo-pallini replied:
Was it considered to add to the pre-commit checks also a check for print statements?

ricardoV94 replied:
@MarcoGorelli suggested a way this could be done on the slack, following your suggestion. Would either one of you be interested in implementing this?

MarcoGorelli replied:
I'm gonna be pretty busy in the next two weeks, but this script should work as a local hook:

```python
import ast
import sys

class Visitor(ast.NodeVisitor):
    def __init__(self, file):
        self.file = file
    def visit_Call(self, node: ast.Call) -> None:
        if isinstance(node.func, ast.Name) and node.func.id == 'print':
            sys.stdout.write(f'{self.file}:{node.lineno}:{node.col_offset} found print statement\n')
            sys.exit(1)

if __name__ == '__main__':
    for file in sys.argv[1:]:
        with open(file) as fd:
            content = fd.read()
        tree = ast.parse(content)
        visitor = Visitor(file)
        visitor.visit(tree)
```

@drabbit17 if you wanted to submit this as a separate PR, I'll review it

matteo-pallini replied:
Happy to give it a stab

ricardoV94 commented:
> Is `_defaultval` deprecated? I haven't been able to find any use of it

Kind of, we still need to refactor the testval/ initialization point logic for V4 as discussed in #4567

ricardoV94 commented on pymc3/distributions/continuous.py:
```diff
@@ -723,42 +686,69 @@ def logp(self, value):
+    lower = (
+        at.as_tensor_variable(floatX(lower))
+        if lower is not None
+        else at.as_tensor_variable(-np.inf)
```
Are these `infs` used anywhere?

matteo-pallini replied:
I guess they would be used when either `lower` or `upper` is None. There were a couple of tests for the cases in which either `upper` or `lower` were None. ie `TestTruncatedNormalLower` and `TestTruncatedNormalUpper`

ricardoV94 replied:
It seems to me they are not used, because the logp changes based on this and doesn't include the `None` extremes anyway. I am not even sure we need to separate between `lower` and `lower_check`.

ricardoV94 commented:
> Is it find to pass transform as argument in dist as per TruncatedNormal?

I am not sure what you mean. The BoundedContinuous class takes care of default transforms. It is possible for a user to specify another transform which will overwrite the default one. The `TruncatedNormal.dist()` initialization, on the other hand, will not do anything with transforms, they only matter for normal distributions intialized within a model.

I am not sure which of these you are referring to.

ricardoV94 commented:
> Need to investigate why pymc3.tests.test_model.TestValueGradFunction.test_aesara_switch_broadcast_edge_cases_2 is failing

That test probably needs to be slightly refactored for V4. The dlogp call might need to be tweaked. What error are you seeing?

matteo-pallini commented:
`I am not sure what you mean. The BoundedContinuous ...`
Sorry, the bullet point wasn't really helpful. I wrote it as a reference for myself and didn't consider how cryptic it was. I was referring to the 2nd case you mentioned. 

I thought that the `transform` argument if passed to `dist` could have reached the `rv_op` call in `Distribution.dist` through `**kwargs` and that the logic to handle `transform` had been moved to aesara. But actually that would have not worked python-wise. So, I guess that the `transform` logic (and argument) will be simply removed with the refactoring.

`The dlogp call might need to be tweaked. What error are you seeing?`
`m.dlogp([mu])({"mu": 0})` is consistently 0. I need to familiarize with the internals of `Model` and `logp`/`dlogp` before being able to tweak the test appropriately. Thanks for letting me know that changing the test is an option worth considering

ricardoV94 commented:
> I thought that the transform argument if passed to dist could have reached the rv_op call in Distribution.dist through **kwargs and that the logic to handle transform had been moved to aesara. But actually that would have not worked python-wise. So, I guess that the transform logic (and argument) will be simply removed with the refactoring.

Only the first argument (the list of parameters) and `size/shape` are ever passed to the rv_op. The other kwargs are intercepted in `Distribution.__new__()` to be used there or forwarded to `Model.register_rv()`

ricardoV94 committed:
Fix failing test case

ricardoV94 commented:
@drabbit17 I pushed a tiny change for the failing test. The issue was that we were passing the RandomVariable to the dlogp function instead of the logp "value" variable. It is an expected V3->V4 refactoring

ricardoV94 commented:
Hi @drabbit17, any progress on this PR?

We merged `V4` into `main` so you will have to redirect the target of this PR.

Let me know if you need any help.

matteo-pallini commented:
> We merged V4 into main so you will have to redirect the target of this PR.

Thanks for the update and congrats!

> Hi @drabbit17, any progress on this PR?

Sorry for the lack of progress here. During the last 2-3 weeks work left me with little/no mental bandwidth. I should be able to pick it up again on the weekend. I would like to, but please feel free to re-assign the issue to someone else in case I am being a blocker, or someone else is keen to pick it earlier than that.

ricardoV94 commented:
@drabbit17 There is no rush, just wanted to check what was your status.

Also, do I understand correctly from the PR title that you intended to refactor the `Wald` distribution?

matteo-pallini commented:
> you intended to refactor the Wald distribution?

Yes

matteo-pallini committed:
Address review feedback

- remove upper and lower checks and check for lack of bounds in logp and
  intervals directly
- pass default value for dist as `testval` to Distribution dist method

ricardoV94 commented on pymc3/distributions/continuous.py:
```diff
@@ -632,101 +655,55 @@ class TruncatedNormal(BoundedContinuous):
+        else:
+            initval = (lower + upper) / 2
+
+        lower = (
```
We shouldn't transform away from `None` unless it is needed, otherwise the branch logic in the logp (and also in the interval transform) is in vain, as it relies on a `None` check. Was this giving you problems?

In the `rng_fn` we can add extra logic to check for `None` and convert to `infs` if needed. If anything, in the `dist` method we could check for `np.inf` and convert those to `None`.

matteo-pallini replied:
Sorry for that. I still have a partial understanding of the whole picture. Haven't had a chance yet to go through the sampling/inference part. So, I still don't know how `logp` is actually used (ie my guess from you comment is that it takes as arguments the dist_params that have already been processed by `dist`, ie the RandomVariable node inputs).

If the `None` are not handled `make_node` in `RandomVariable` will fail, as it's expecting objects that could be converted to tensor variables. So, I don't think that moving that check to `rng_fn` would solve the problem
https://github.com/aesara-devs/aesara/blob/3acdd78e7a5d27e70cd8de3fc00bd5f010c65ee3/aesara/tensor/random/op.py#L360

Stack trace
```
self = <pymc3.tests.test_distributions_random.TestTruncatedNormalLowerTau object at 0x7fb50c50a3d0>

    def test_distribution(self):
        self.validate_tests_list()
>       self._instantiate_pymc_rv()

pymc3/tests/test_distributions_random.py:288: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pymc3/tests/test_distributions_random.py:298: in _instantiate_pymc_rv
    self.pymc_rv = self.pymc_dist.dist(
pymc3/distributions/continuous.py:709: in dist
    res = super().dist([mu, sigma, lower, upper], testval=initval, **kwargs)
pymc3/distributions/distribution.py:285: in dist
    rv_out = cls.rv_op(*dist_params, size=create_size, **kwargs)
../../miniconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/random/op.py:323: in __call__
    res = super().__call__(rng, size, dtype, *args, **kwargs)
../../miniconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/graph/op.py:271: in __call__
    node = self.make_node(*inputs, **kwargs)
../../miniconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/random/op.py:360: in make_node
    dist_params = tuple(
../../miniconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/random/op.py:361: in <genexpr>
    as_tensor_variable(p) if not isinstance(p, Variable) else p
../../miniconda3/envs/pymc3-dev-py38/lib/python3.8/site-packages/aesara/tensor/__init__.py:41: in as_tensor_variable
    return _as_tensor_variable(x, name, ndim, **kwargs)
../../miniconda3/envs/pymc3-dev-py38/lib/python3.8/functools.py:875: in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
```

matteo-pallini replied:
I wonder whether checking for `inf` and `-inf` rather than `None` in `logp` and in the interval transform could be a feasible approach

ricardoV94 replied:
That's spot on. I didn't see it immediately.

One ugly solution is to have 3 versions of the distribution (each with its own transform and logp / random methods) but I would rather not.

Another is to wrap the infinity arguments in an `at.constant` which we can safely query (I think) when calling the logp method (first checking if it's a constant and then if it's +- infinity). 

For the interval transform we can define the different variations at creation time in the `new` or `dist` classmethod (not sure which).

ricardoV94 replied:
Seems like you got ahead of me!

ricardoV94 replied:
> I wonder whether checking for `inf` and `-inf` rather than `None` in `logp` and in the interval transform could be a feasible approach

Yeah we can but only within an `at.constant` otherwise we would need symbolic branching with `at.switch` instead of the python `if else`. Since the point is to return a more optimized graph I would go with the constant + python branching.

ricardoV94 replied:
Do we have tests that cover this specific `+-inf` behavior?

matteo-pallini replied:
`TestTruncatedNormalLowerTau` and `TestTruncatedNormalUpperTau`

ricardoV94 replied:
I mean testing the interval and the logp is actually doing what it should (returning different expressions based on whether one or both of the inputs are `None`)

matteo-pallini replied:
I added `pymc3.tests.test_distributions.TestBoundedContinuous` to test the interval. I haven't been able to test `BoundedContinuous` directly, so I had to do so through a `TruncatedNormal`

I also added a slightly more convoluted test (which seems sensible) for the interval behavior `pymc3.tests.test_distributions.TestBoundedContinuous.test_missing_upper_bound_with_richer_context` which is failing. I need to investigate why

I added a couple of tests for the logp with unbounded intervals `pymc3.tests.test_distributions.TestMatchesScipy.test_truncated_normal`.

Let me know if you think it makes sense to add more

matteo-pallini committed:
Refactor Wald distribution

matteo-pallini commented on pymc3/tests/test_distributions_random.py:
```diff
@@ -525,6 +500,147 @@ def seeded_kumaraswamy_rng_fn(self):
+    ]
+
+
+class TestWaldPureScipy(BaseTestDistribution):
```
I would expect this to pass, but it doesn't. I wonder whether it fails because somehow the random sampler seeding is not returning the expected results when we have more than one distribution being sampled in `rng_fn`. In this case both `rng.normal(size=size)` and `rng.uniform(size=size)`

ricardoV94 replied:
If the implementation is not the same, it's not necessarily the case they would match. Maybe the order the rng methods are called in also matters? By the way, is there a reason why we can't simply delegate to the `scipy` method in the new `RandomVariable`?

matteo-pallini replied:
> Maybe the order the rng methods are called in also matters

yeah I think that in that case the order of the rng mattered. I think that the Scipy implementation of the Wald distribution doesn't match the PyMC one though. The Numpy one seems to be right though, so I replace the custom sampling with the numpy one

https://numpy.org/doc/stable/reference/random/generated/numpy.random.wald.html

ricardoV94 commented on pymc3/distributions/continuous.py:
```diff
@@ -162,8 +162,16 @@ def transform_params(rv_var):
-            upper = at.as_tensor_variable(upper) if upper is not None else None
+            lower = (
+                None
+                if lower is None or (isinstance(lower, TensorConstant) and lower.value == -np.inf)
```
I think this will raise an error if you get a non scalar constant in other contexts

matteo-pallini replied:
Good point! I added a test for that, which is currently failing `pymc3.tests.test_distributions.TestBoundedContinuous.test_missing_upper_bound_array` still need to fix it though, I should be able to do it later today. I guess it will be a matter of having a check for scalar and one for arrays. btw I should also create an array of np.infs within `dist` when the other params are arrays.

ricardoV94 replied:
By the way, I have been working on something related for Censored distributions, and this is how I wrote the "constant inf check" so far: https://github.com/aesara-devs/aeppl/blob/46fac1347d129e0a6c7391c87bbae456f51aa009/aeppl/truncation.py#L36 (related to https://github.com/aesara-devs/aeppl/pull/22)

I am not very happy with the solution. I am still thinking whether it makes more sense to subclass (in my case the Censored class), in your case the Truncated Normal, into left-truncated, right-truncated, and truncated, depending on the bounds.

matteo-pallini replied:
Thanks for sharing the link, I will give it a look later.

Possibly it may just be as simple as doing as per below:
```python
lower = (
      at.as_tensor_variable(lower)
      if lower is not None
      and (isinstance(lower, TensorConstant) and -np.inf not in lower.value)
      else None
  )
```

I pushed an updated (and not too ugly) approach that is passing the test with arrays. I need to check it more thoroughly though

ricardoV94 replied:
Can't read that code in my mind right now, but I am thinking of the case where you have lower = [-np.inf, 0], i.e., when one of the replications of the variable is lower bounded and the other is not, in which case the logp and transform should treat the variable as being lowerbounded. Not sure if your snippet would do the right thing.

matteo-pallini replied:
Good point, I only realized it after having shared the snippet. The approach with `np.all` is pretty neat! thanks for sharing it

matteo-pallini committed:
Add logp and transform tests for truncated normal

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -2707,6 +2723,65 @@ def test_bound():
         BoundPoissonPositionalArgs = Bound(Poisson, upper=6)("x", 2.0)
 
 
+class TestBoundedContinuous:
```
Tests look like they go in the right direction. I'll check them more carefully later, now I just wanted to mention a nitpick: I would pass always the missing lower / upper explicitly as `None`, to make them easier to read.

matteo-pallini committed:
Move Wald distribution to use the numpy rvs implementation

matteo-pallini commented:
One test still failing, everything else should be close to be finalised (as far as I can tell)

ricardoV94 commented on pymc3/distributions/continuous.py:
```diff
@@ -632,99 +673,63 @@ class TruncatedNormal(BoundedContinuous):
-        else:
-            self._defaultval = (self.lower + self.upper) / 2
-
+        single_param_length = 0 if not sigma.shape else sigma.shape[0]
```
The shape could be determined by `mu` or even the other `bound` as well.

Also, the shape may be symbolic (if you have a symbolic input). I am not sure if `np.repeat` is "overloaded" by Aesara if the shape is symbolic, but there should be an `at.repeat` equivalent that you can safely use.

matteo-pallini replied:
I played a bit with different options and I cannot find a decent one supporting symbolic shapes.

If use a python if else I will be able to return a scalar when scalars as passed as inputs and an array otherwise. However I won't be able to use the symbolic shape (unless I eval it). eg:
```python 
def _handle_missing_bound(bound_value, expected_input_length, replacement_value): 
	if bound_value is not None: 
    	    return at.as_tensor_variable(floatX(bound_value)) 
	elif expected_input_length == 0: 
    	    return at.as_tensor_variable(floatX(replacement_value)) 
	else: 
    	    return at.repeat(at.as_tensor_variable(replacement_value), at.as_tensor_variable(expected_input_length))

```

If on the other hand I use at,switch I end up always having an array as output (an empy one when mu_shape is empty, ie the input is a scalar)
```python
switch(at.gt(mu_shape.size, at.as_tensor_variable(0)), at.repeat(at.as_tensor_variable(floatX(replacement_value)), mu_shape.size), replacement_value)
```

Alternatively I can use at.repeat directly, but again I end up always having an array as output
```python
repeat = at.switch(at.gt(mu_shape.size, at.as_tensor_variable(0)), mu_shape.shape[0], 1)
bound_value = at.repeat(at.as_tensor_variable(floatX(replacement_value)), repeat).eval()
```
None of the cases above seems a serious solution to me, at the same time I cannot think a decent symbolic approach supporting both scalars and arrays. I have fairly little Theano knowledge, so there may be an obvious solution to this

btw I am also going to add a check in `test_distributions_random` checking that the parametrization is as expected when symbolic inputs are passed to `dist`

ricardoV94 replied:
I might have missed / forgot it, but why can't we pass a simple constant `-np.inf`. The RandomVariable should be able to broadcast it to a vector when needed.

matteo-pallini replied:
Thanks for stepping in and improving it.

> The RandomVariable should be able to broadcast it to a vector when needed.

Oh! I didn't know about it and I thought it was necessary to handle it in`.dist`. That's the reason why I wanted to use `np.repeat` /`at.repeat`

ricardoV94 commented on pymc3/distributions/continuous.py:
```diff
@@ -900,6 +897,21 @@ def _distr_parameters_for_repr(self):
+
+    @classmethod
+    def rng_fn(cls, rng, mu, lam, alpha, size):
+        return getattr(np.random.RandomState, cls.name)(rng, mu, lam, size) + alpha
```
```suggestion
        rng.wald(mu, lam, size=size) + alpha
```
Does this work?

ricardoV94 commented on pymc3/distributions/continuous.py:
```diff
@@ -632,99 +673,63 @@ class TruncatedNormal(BoundedContinuous):
+        elif lower is not None and upper is None:
+            initval = lower + 1.0
+        else:
+            initval = (lower + upper) / 2
```
This can be commented out for now

ricardoV94 commented on pymc3/distributions/continuous.py:
```diff
@@ -738,38 +743,30 @@ def logp(self, value):
-        norm = self._normalization()
-        logp = Normal.dist(mu=mu, sigma=sigma).logp(value) - norm
+        unbounded_lower = isinstance(lower, TensorConstant) and lower.value == -np.inf
+        unbounded_upper = isinstance(upper, TensorConstant) and upper.value == np.inf
```
Needs the same logic as above with the `np.all`

ricardoV94 committed:
Simplify None bound replacement

ricardoV94 commented:
I pushed some changes, I think the logic works fine with a scalar `-+np.inf` instead of `lower` / `upper`. I also rebased on `main` and fixed some merge conflicts that emerged. Don't forget to rebase from here before pushing new changes!

ricardoV94 committed:
Make `TestBoundedContinuous` more readable

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -2714,6 +2725,104 @@ def test_bound():
+        assert lower_interval.value == -1
+        assert upper_interval is None
+
+    def test_rich_context(self):
```
I think we can remove this one. Does not seem to test anything extra

matteo-pallini replied:
The error trace you shared below is coming from that test, Originally I left it because I was expecting it to pass and it didn't, so I wanted to investigate why. Judging by [this thread](https://stackoverflow.com/questions/40845304/runtimewarning-numpy-dtype-size-changed-may-indicate-binary-incompatibility) I would say that the warning is unrelated.

The error seems to be coming from `pymc3.model.Model.set_initval`. When running the Truncated with lower=None there is a mismatch in the number of dimensions between the `rv_value_var` (which is a scalar) and the `initval` (which is an array). The second is generated by `initval_fn(), but its ndim is due to `transform.forward(rv_var, value)`, I think. So, it may be possible that the interval is returning a wrong value due to the refactoring. But I haven't been able to replicate the issue with simpler tests, so I may be simply wrong. I may have written the test incorrectly, or there may be a bug. I think it's worth keeping the test at least until we cannot make it pass. For now I removed it though

ricardoV94 replied:
I also saw that thread about not being harmful... but it's suspicious that the warning appears just after the failure, which comes from an Aesara check that ndims did not change and the warning does not appear in the other successful runs of the same test.

ricardoV94 commented:
Seeing a weird warning in the failing jobs: https://github.com/pymc-devs/pymc3/pull/4711/checks?check_run_id=2950387511#step:7:2342

>  /usr/share/miniconda/envs/pymc3-dev-py37/lib/python3.7/importlib/_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 192 from C header, got 216 from PyObject


Which seems to be related to the failure just before: https://github.com/pymc-devs/pymc3/pull/4711/checks?check_run_id=2951980417#step:7:2332

```python
          if self.ndim != data.ndim:
            raise TypeError(
>               f"Wrong number of dimensions: expected {self.ndim},"
                f" got {data.ndim} with shape {data.shape}."
            )
E           TypeError: Wrong number of dimensions: expected 0, got 1 with shape (1,).
```

matteo-pallini committed:
Add arrays test for Wald, remove useless test and simplify sampling from numpy wald

ricardoV94 committed:
Remove redundant tests

ricardoV94 commented:
@drabbit17 check my [last commit](https://github.com/pymc-devs/pymc3/pull/4711/commits/26a442d5cab641ab0b82e673434eedc0e04b461e). I removed some tests (or `tests_to_run`) that felt unnecessary. Let me know if you disagree. Otherwise I think this PR is ready to merge.

I tested creating a bunch of TruncatedNormals with random sizes and lower/ upper parameters in a single model and I did not find any issues with the `initval` like the one we were getting in that `rich_context` model. So I am pretty confident that it was not an issue on our side.

matteo-pallini committed:
remove unnecessary variable assignment

matteo-pallini commented:
> Otherwise I think this PR is ready to merge

Agreed, I did a very small change. sorry for dragging the PR for so long and thanks for the support (it would have been way faster for you to simply do the whole thing yourself :-) ).

> So I am pretty confident that it was not an issue on our side.

our side as this PR or PyMC?

I have been trying to replicate the test failure for `pymc3/tests/test_distributions_random.py::TestNestedRandom::test_TruncatedNormal` locally but I haven't been able to, can you? (the same goes for the `rich_context` test). Are the github tests being ran through a docker container? if so, is it possible to download it?

ricardoV94 commented:
> Agreed, I did a very small change. sorry for dragging the PR for so long and thanks for the support (it would have been way faster for you to simply do the whole thing yourself :-) ).

I disagree. The dynamic interval thing is something that we needed to figure out, and will be used in other places as well. It was really great that you dived in and started figuring it out.

> > So I am pretty confident that it was not an issue on our side.
> 
> our side as this PR or PyMC?

Both. I am pretty confident it was an issue with incompatible numpy / aesara binaries, that emerged on that specific environment.

> I have been trying to replicate the test failure for `pymc3/tests/test_distributions_random.py::TestNestedRandom::test_TruncatedNormal` locally but I haven't been able to, can you? (the same goes for the `rich_context` test). Are the github tests being ran through a docker container? if so, is it possible to download it?

The `TestNestedRandom` seems like a weird caching issue, because we recently made tests marked as xfail strict (the tests fail if the test pass) and in this case it seems to still be surprised that the tests are passing. https://github.com/pymc-devs/pymc3/pull/4711/checks?check_run_id=2973646757#step:8:682

I can't find a explicit error message in the logs related to the "failing tests" (@michaelosthege any ideas?)

michaelosthege commented:
I added the `mark.xfail` on `test_TruncatedNormal` because the test used a nonexistent API, but the distribution was not yet refactored.
This PR refactored the distribution, so now the test is passing.
👉 Just remove the xfail.

michaelosthege committed:
Remove xfail mark on TrucatedNormal tests

michaelosthege commented:
Rebased and removed that `mark.xfail`. Should go green now, but please check if my rebase didn't mess up anything :)

ricardoV94 commented:
> I added the `mark.xfail` on `test_TruncatedNormal` because the test used a nonexistent API, but the distribution was not yet refactored.
> This PR refactored the distribution, so now the test is passing.
> 👉 Just remove the xfail.


I was convinced that I had removed it xD

michaelosthege commented:
@ricardoV94 the rebase was tricky too. You probably ended up with another tests `mark.xfail` earlier.

Both test failures seem to be related to the known problem with the `find_MAP` #4771.
Merge away?

ricardoV94 commented:
> @ricardoV94 the rebase was tricky too. You probably ended up with another tests `mark.xfail` earlier.
> 
> Both test failures seem to be related to the known problem with the `find_MAP` #4771.
> Merge away?

The MLE one will be adjusted in #4833. ~~Let me check the other one quickly~~ Yeah the second one looks another MAP issue. Does not seem related to any changes in this PR

ricardoV94 commented:
Great work @drabbit17 This was a fun one to crack. Looking forward to your next PR :)
